src/gluonts/dataset/rolling_dataset.py

```diff
@@ -0,0 +1,177 @@
+    been rolled upon based on the parameters supplied. Below follows an 
+    explanation and examples of how the different parameters can be used to generate
+    differently rolled datasets.
+    
+    The 'rolling' will happen on the data available in the provided window between the 
+    start_time and the end_time for each timeseries. If end_time is omitted, rolling
+    happens on all datapoints from start_time until the end of the timeseries. 
+    The way the data is rolled is governed by the strategy used. 
+
+    Below examples will be based on this one timeseries long dataset
+    [{
+    target=[1,2,3,4,5,6,7,8,9,10,11,12,13,14,15]
+    start='2000-1-1-01'
+    }]
+    
+    applying generate_rolling_datasets on this dataset with:
+
+    start_time = pd.Timestamp('2000-1-1-06', '1H')
+    end_time = pd.Timestamp('2000-1-1-10', '1H')
+    strategy = StepStrategy(prediction_length=2)
+
+    returns a new dataset as follows (only target values shown for brevity):
+
+    [1, 2, 3, 4, 5, 6, 7, 8, 9, 10],
+    [1, 2, 3, 4, 5, 6, 7, 8, 9],
+    [1, 2, 3, 4, 5, 6, 7, 8],
+    [1, 2, 3, 4, 5, 6, 7]
+
+    i.e. maximum amount of rolls possible between the end_time and start_time.
+    The basic_strategy only cuts the last value of the target for as long as
+    there is enough values after start_time to perform predictions on.
+
+    When no end time is provided the output is as below since all datapoints
+    from start_time will be rolled over.
+
+    [1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15]
+    [1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14]
+    [1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13]
+    [1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12]
+    [1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11]
+    [1, 2, 3, 4, 5, 6, 7, 8, 9, 10]
+    [1, 2, 3, 4, 5, 6, 7, 8, 9]
+    [1, 2, 3, 4, 5, 6, 7, 8]
+    [1, 2, 3, 4, 5, 6, 7]
+
+    One can change the step_size of the strategy as below:
+
+    strategy = StepStrategy(prediction_length=2, step_size=2)
+
+    This causes fewer values to be in the output which, 
+    when prediction_length matches step_size, ensures that each prediction
+    will be done on unique/new data. Below is the output of this run.
+    
+    [1, 2, 3, 4, 5, 6, 7, 8, 9, 10]
+    [1, 2, 3, 4, 5, 6, 7, 8]
+
+    Not setting an end time and using the step_size=2 results in
+    the below dataset.
+
+    [1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15]
+    [1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13]
+    [1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11]
+    [1, 2, 3, 4, 5, 6, 7, 8, 9]
+    [1, 2, 3, 4, 5, 6, 7]
+    """
+    assert dataset, "a dataset to perform rolling evaluation on is needed"
+    assert start_time, "a pandas Timestamp object is needed for the start time"
+    assert strategy, """a strategy to use when rolling is needed, for example
+        gluonts.dataset.rolling_dataset.StepStrategy"""
+    if end_time:
+        assert end_time > start_time, "end time has to be after the start time"
+
+    ds = []
```

Comment:
Sorry to jump in late, not sure this was already discussed: does this need to be a list? Couldn't we use a generator here?

Reply:
I think we at least discussed this offline.

The issue is that we need to run this twice, when using `backtest.make_evaluation_predictions`. But maybe that's an issue of `make_evaluation_predictions`